slatedb/src/record/store.rs

```diff
@@ -0,0 +1,365 @@
+
+    pub(crate) async fn write(&self, id: u64, value: &T) -> Result<(), SlateDBError> {
+        let path = self.path_for(id);
+        self.object_store
+            .put_if_not_exists(&path, self.codec.encode(value))
+            .await
+            .map_err(|err| {
+                if let AlreadyExists { path: _, source: _ } = err {
+                    SlateDBError::FileVersionExists
+                } else {
+                    SlateDBError::from(err)
+                }
+            })?;
+        Ok(())
+    }
+
+    pub(crate) async fn try_read(&self, id: u64) -> Result<Option<T>, SlateDBError> {
+        let path = self.path_for(id);
+        match self.object_store.get(&path).await {
+            Ok(obj) => match obj.bytes().await {
+                Ok(bytes) => self.codec.decode(&bytes).map(Some),
+                Err(e) => Err(SlateDBError::from(e)),
+            },
+            Err(e) => match e {
+                Error::NotFound { .. } => Ok(None),
+                _ => Err(SlateDBError::from(e)),
+            },
+        }
+    }
+
+    pub(crate) async fn try_read_latest(&self) -> Result<Option<(u64, T)>, SlateDBError> {
+        let files = self.list(..).await?;
+        if let Some(file) = files.last() {
+            return self
+                .try_read(file.id)
+                .await
+                .map(|opt| opt.map(|v| (file.id, v)));
+        }
+        Ok(None)
+    }
+
+    // List files for this record type within an id range
+    pub(crate) async fn list<R: RangeBounds<u64>>(
+        &self,
+        id_range: R,
+    ) -> Result<Vec<GenericFileMetadata>, SlateDBError> {
+        let base = &Path::from("/");
+        let mut files_stream = self.object_store.list(Some(base));
+        let mut items = Vec::new();
+        while let Some(file) = match files_stream.next().await.transpose() {
+            Ok(file) => file,
+            Err(e) => return Err(SlateDBError::from(e)),
+        } {
+            match self.parse_id(&file.location) {
+                Ok(id) if id_range.contains(&id) => {
+                    items.push(GenericFileMetadata {
+                        id,
+                        location: file.location,
+                        last_modified: file.last_modified,
+                        size: file.size as u32,
+                    });
+                }
+                Err(_) => warn!(
+                    "unknown file in record directory [location={:?}]",
+                    file.location
+                ),
+                _ => {}
+            }
+        }
+        items.sort_by_key(|m| m.id);
+        Ok(items)
+    }
+
+    // Delete a specific versioned file (no additional validation)
+    pub(crate) async fn delete(&self, id: u64) -> Result<(), SlateDBError> {
+        let path = self.path_for(id);
+        debug!("deleting record [record_path={}]", path);
+        self.object_store
+            .delete(&path)
+            .await
+            .map_err(SlateDBError::from)
+    }
+}
```

Comment:
Some tests would be good. Can probably vibe code them given that we have a lot of manifest store tests already.